File: Engine/src/main/java/org/etieskrill/engine/input/controller/KeyCharacterController.java
package org.etieskrill.engine.input.controller;

import org.etieskrill.engine.input.Input;
import org.etieskrill.engine.input.InputBinding;
import org.etieskrill.engine.input.KeyInputManager;
import org.etieskrill.engine.input.Keys;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.joml.Vector3f;

import java.util.function.Supplier;

/**
 * @param <T> the type of object being controlled
 */
public class KeyCharacterController<T> extends KeyInputManager {

    protected final T target;
    protected float speed = 1;
    protected @Nullable Supplier<Boolean> updateCondition;
    protected final Updatable<T> updateFunction;

    protected final Vector3f deltaPosition;

    public KeyCharacterController(T target, Updatable<T> updateFunction) {
        this.target = target;
        this.updateFunction = updateFunction;

        this.deltaPosition = new Vector3f();

        addBindings(Input.bind(Keys.W).on(InputBinding.Trigger.PRESSED).to(() -> deltaPosition.add(0, 0, 1)));
        addBindings(Input.bind(Keys.S).on(InputBinding.Trigger.PRESSED).to(() -> deltaPosition.add(0, 0, -1)));
        addBindings(Input.bind(Keys.A).on(InputBinding.Trigger.PRESSED).to(() -> deltaPosition.add(1, 0, 0)));
        addBindings(Input.bind(Keys.D).on(InputBinding.Trigger.PRESSED).to(() -> deltaPosition.add(-1, 0, 0)));
        addBindings(Input.bind(Keys.SPACE).on(InputBinding.Trigger.PRESSED).to(() -> deltaPosition.add(0, 1, 0)));
        addBindings(Input.bind(Keys.SHIFT).on(InputBinding.Trigger.PRESSED).to(() -> deltaPosition.add(0, -1, 0)));
    }

    @Override
    public void update(double delta) {
        updateInputManager(delta);
        if (updateCondition == null || updateCondition.get()) {
            updateFunction.update(delta, target, deltaPosition, speed);
        }
        deltaPosition.zero();
    }

    @FunctionalInterface
    public interface Updatable<T> {
        void update(double delta, T target, Vector3f deltaPosition, float speed);
    }

    protected void updateInputManager(double delta) {
        super.update(delta);
    }

    public float getSpeed() {
        return speed;
    }

    public KeyCharacterController<T> setSpeed(float speed) {
        this.speed = speed;
        return this;
    }

    public KeyCharacterController<T> setUpdateCondition(@NotNull Supplier<Boolean> updateCondition) {
        this.updateCondition = updateCondition;
        return this;
    }

}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/input/controller/KeyCharacterController.kt
package org.etieskrill.engine.input.controller

import org.etieskrill.engine.input.InputTriggerEdge
import org.etieskrill.engine.input.Key
import org.etieskrill.engine.input.KeyInputManager
import org.etieskrill.engine.input.bindKey
import org.joml.Vector3f

/**
 * @param T the type of object being controlled
 */
open class KeyCharacterController<T>(
    protected val target: T,
    protected val updateFunction: Updatable<T>,

    var speed: Float = 1f,
    var updateCondition: (() -> Boolean)? = null,
) : KeyInputManager() {

    protected val deltaPosition = Vector3f()

    init {
        addBindings(
            bindKey(Key.W, on = InputTriggerEdge.PRESSED) { deltaPosition.add(0f, 0f, 1f) },
            bindKey(Key.S, on = InputTriggerEdge.PRESSED) { deltaPosition.add(0f, 0f, -1f) },
            bindKey(Key.A, on = InputTriggerEdge.PRESSED) { deltaPosition.add(-1f, 0f, 0f) },
            bindKey(Key.D, on = InputTriggerEdge.PRESSED) { deltaPosition.add(1f, 0f, 0f) },
            bindKey(Key.SPACE, on = InputTriggerEdge.PRESSED) { deltaPosition.add(0f, 1f, 0f) },
            bindKey(Key.SHIFT, on = InputTriggerEdge.PRESSED) { deltaPosition.add(0f, -1f, 0f) }
        )
    }

    fun interface Updatable<T> {
        fun update(delta: Double, target: T, deltaPosition: Vector3f, speed: Float)
    }

    override fun update(delta: Double) {
        super.update(delta)
        if (updateCondition?.invoke() ?: true) {
            updateFunction.update(delta, target, deltaPosition, speed)
        }
        deltaPosition.zero()
    }

}
